docs/specs/code/SIGNING.md

```diff
@@ -0,0 +1,424 @@
+
+```
+COSE_Encrypt0 = [
+    Headers,
+    ciphertext : bstr / nil,
+]
+
+PasswordEncryption = 16 (COSE_Encrypt0)
+```
+
+The `Headers` MUST have `enc_type: 1` in the `unprotected` field.
+
+The `ciphertext` is simply the result of [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md) on the `signed_message`.
+
+#### Public key based encryption
+
+We only allow encrypting based on `ED25519 extended` public keys (the ones used for Cardano). To encrypt based on these public keys, you must
+
+1) Compute a password consisting of 22 case-sensitive alphanumeric (a–z, A–Z, 0–9) characters (this gives you ~128 bits of entropy)
+
+Now, for each receiver, you must
+
+2) Compute an ephemeral key pair using `ED255519 extended`
+1) Compute the shared DH secret between the private key from step (1) and the public key received (using the `exchange` functionality)
+1) Compute the `Blake2b512` hash of the secret
+1) Use this as the password to encrypt the password in (1) using [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md)
+
+The structure will look like the following:
+
+```
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, # contained signed_message encrypted with random password
+  recipients : [+COSE_recipient]
+]
+
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains random password encrypted with shared secret
+]
+
+PubKeyEncryption = 96 (COSE_Encrypt)
+```
+
+The `Headers` for the body MUST have `enc_type: 2` in the `unprotected` field.
+
+The `Headers` for the recipient MUST have a `eph_pub: bstr` label containing the public key of the ephemeral keypair.
+
+## Multisig
+
+TODO
+
+## More headers
+
+### CWT
+
+CWT tags MAY be introduced inside `protected` headers as a CBOR map with the key `cwt`.
+
+### Payload encoding
+
+To solve `E3`, `signed_message` body header MUST `is_ascii: bool` as a `protected` header which defined whether or not the bytes the `payload` should be considered as ASCII .
+
+If the `payload` is not ASCII, hardware wallets MAY instead show the `Blake2b244` hash of the `payload` to users. `244` is used to produce a smaller hash (easier to verify on small screen).
+
+## User-Facing Encoding
+
+Once we have our top-level `encrypted_message` or `signed_message` we need to encode them in way that can be displayed to users.
+
+We define the encoding in three parts `prefix || data || checksum`.
+
+### Prefix
+
+We need a human-readable prefix. We use "CM" for "Cardano Message" followed by the message type:
+
+- `encrypted_message`: `cme_`
+- `signed_message`: `cms_`
+- `mac_message`: `cmm_` (unused in this spec)
+
+### Data
+
+Data is simply the `base64url` encoding of the message
+
+### Checksum
+
+We use `CRC32` on the data for the checksum 
```

Comment:
I used `CRC32` since  we use it elsewhere but it seems CRC in general has fallen out of popularity. I'm open for a different checksum method if we want.